### PR TITLE
(CDAP-7008) Added a DefaultPrivilegesManager that invalidates both th…

### PR DESCRIPTION
**Pull request by bdmogal**
…e proxy and the enforcement caches when privileges are updated.

Jira: [CDAP-7008](https://issues.cask.co/browse/CDAP-7008)
Build: http://builds.cask.co/browse/CDAP-RUT77

### CODE REVIEW

**Comment by gokulavasan** [comment]
@bdmogal Will this solve the issue reported in the JIRA? Or does it just reduce the time window in which that error can occur?

**Comment by bdmogal** [comment]
It resolved the app deployment scenario for me. I created namespaces and instantaneously deployed apps on them, and that succeeded. However, I'm not sure it covers invalidation in general across CDAP. Will have to fix all those scenarios as we see them.

**Comment by anwar6953** [comment]
it seems like it will solve the issue for caching in master, which is most of it.
Won't solve it in stream container or explore container, etc.

**Comment by gokulavasan** [comment]
@bdmogal Solving caching in master is what the JIRA requires I guess. That answers my question. Thanks!

**Comment by bdmogal** [comment]
That's correct @anwar6953. For streams/explore containers, I'm not yet sure. Over there the strategy may have to be that for certain specific calls, we pass a parameter to bypass caching. But will fix these guys one by one as we see it. I'm even ok documenting that behavior for stream and explore containers for 3.5.

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java [hunk 1]
```diff
@@ -176,6 +178,18 @@ protected void shutDown() throws Exception {
     return cacheEnabled ? authPolicyCache.get(principal) : fetchPrivileges(principal);
   }
 
+  protected void doInvalidate(final Predicate<Principal> predicate) {
+    Iterable<Principal> filtered = Iterables.filter(authPolicyCache.asMap().keySet(), new Predicate<Principal>() {
+      @Override
+      public boolean apply(Principal input) {
+        return predicate.apply(input);
```
Comment (chtyim) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java]: Why need to new a `Predicate` and just calling the other? Can't you just use the `predicate` in the argument?

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java [hunk 2]
```diff
@@ -176,6 +178,18 @@ protected void shutDown() throws Exception {
...
+      }
+    });
+    for (Principal principal : filtered) {
```
Comment (chtyim) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java]: Simply call `invalidateAll(filtered)`

Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java]: oops, yes!

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java (hunk 1 above)
Comment (bdmogal) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AbstractAuthorizationService.java]: oops, yes!

**Comment by bdmogal** [comment]
@chtyim addressed initial comments

**Comment by anwar6953** [comment]
lgtm

**Comment by bdmogal** [comment]
Want to test it once more on a cluster. Will merge after that.

**Comment by bdmogal** [comment]
Tested on a cluster. Reported https://issues.cask.co/browse/CDAP-7094 for handling stale privileges in system containers. Merging this.

**Pull request merged**